--- fs/fuse/dir.c.original	2014-11-05 15:01:09.977001397 +0100
+++ fs/fuse/dir.c	2014-11-05 15:02:01.280001398 +0100
@@ -187,7 +187,7 @@
 	if (inode && is_bad_inode(inode))
 		goto invalid;
 	else if (time_before64(fuse_dentry_time(entry), get_jiffies_64()) ||
-		 (nd->flags & LOOKUP_REVAL)) {
+		 (nd && (nd->flags & LOOKUP_REVAL))) {
 		int err;
 		struct fuse_entry_out outarg;
 		struct fuse_req *req;
